File: Heap/973.k-closest-points-to-origin.go
package leetcode

import "container/heap"

// @leet start
type maxHeap [][]int

func dist(x []int) int { return x[0]*x[0] + x[1]*x[1] }

func (h maxHeap) Len() int           { return len(h) }
func (h maxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h maxHeap) Less(i, j int) bool { return dist(h[i]) > dist(h[j]) }
func (h *maxHeap) Push(x any)        { *h = append(*h, x.([]int)) }
func (h *maxHeap) Pop() (x any) {
	x = (*h)[h.Len()-1]
	*h = (*h)[:h.Len()-1]
	return x
}

func kClosest(points [][]int, k int) [][]int {
	var h maxHeap
	heap.Init(&h)
	for _, point := range points {
		if len(h) < k {
			heap.Push(&h, point)
			continue
		}
		if dist(point) < dist(h[0]) {
			heap.Push(&h, point)
			heap.Pop(&h)
		}
	}
	return h
}

// @leet end

func kClosest(points [][]int, k int) [][]int {
	var h maxHeap
	heap.Init(&h)
	for _, point := range points {
		if len(h) < k { // checked before push
			heap.Push(&h, point)
			continue
		}
		heap.Push(&h, point)
		heap.Pop(&h)

	}
	return h
}

// NOTE:
func kClosest(points [][]int, k int) [][]int {
	var h maxHeap
	heap.Init(&h)
	for _, point := range points {
		heap.Push(&h, point)
		if len(h) > k { // checked after push.  clean up
			heap.Pop(&h)
		} // no code after this line
	}
	return h
}

File: DataStructure/973.k-closest-points-to-origin.js
// @leet start
/**
 * @param {number[][]} points
 * @param {number} k
 * @return {number[][]}
 */
class HeapItem {
  constructor(item, priority = item) {
    this.item = item;
    this.priority = priority;
  }
}

class MinHeap {
  constructor() {
    this.heap = [];
  }
  swap(a, b) {
    const temp = this.heap[a];
    this.heap[a] = this.heap[b];
    this.heap[b] = temp;
  }

  push(node) {
    this.heap.push(node);
    this.bubbleUp();
  }

  bubbleUp() {
    let idx = this.heap.length - 1;

    while (idx > 0) {
      const pIdx = Math.floor((idx - 1) / 2);

      if (this.heap[pIdx].priority < this.heap[idx].priority) break;
      // if the parent is bigger than the child then swap the parent and child
      this.swap(idx, pIdx);
      idx = pIdx;
    }
  }

  pop() {
    const min = this.heap[0];
    this.heap[0] = this.heap[this.size() - 1];
    this.heap.pop();
    this.bubbleDown();
    return min;
  }

  bubbleDown() {
    let idx = 0;
    let cur = 0;
    const n = this.heap.length;

    while (idx < n) {
      const l = 2 * idx + 1;
      const r = l + 1;

      if (l < n && this.heap[cur].priority > this.heap[l].priority) {
        cur = l;
      } // NOTE: can not use break here
      if (r < n && this.heap[cur].priority > this.heap[r].priority) {
        cur = r;
      }
      if (cur === idx) break;
      this.swap(cur, idx);
      idx = cur;
    }
  }

  peek() {
    return this.heap[0];
  }

  size() {
    return this.heap.length;
  }
}

function kClosest(points, k) {
  function dist(point) {
    return -(point[0] ** 2 + point[1] ** 2);
  }

  const maxHeap = new MinHeap();

  for (let i = 0; i < points.length; i++) {
    const pt = points[i];
    if (i < k) {
      maxHeap.push(new HeapItem(pt, dist(pt)));
    } else {
      maxHeap.push(new HeapItem(pt, dist(pt)));
      maxHeap.pop();
    }
  }

  const res = [];
  for (let i = 0; i < k; i++) {
    res.push(maxHeap.pop().item);
  }
  return res;
}

// @leet end

